src/data/polynomial/unit_trinomial.lean

```diff
@@ -0,0 +1,315 @@
+  { rwa trinomial_trailing_coeff' hkm hmn },
+  { rwa trinomial_middle_coeff hkm hmn },
+  { rwa trinomial_leading_coeff' hkm hmn },
+end
+
+lemma leading_coeff_is_unit (hp : p.is_unit_trinomial) : is_unit p.leading_coeff :=
+hp.coeff_is_unit (nat_degree_mem_support_of_nonzero hp.ne_zero)
+
+lemma trailing_coeff_is_unit (hp : p.is_unit_trinomial) : is_unit p.trailing_coeff :=
+hp.coeff_is_unit (nat_trailing_degree_mem_support_of_nonzero hp.ne_zero)
+
+end is_unit_trinomial
+
+lemma is_unit_trinomial_iff :
+  p.is_unit_trinomial ↔ p.support.card = 3 ∧ ∀ k ∈ p.support, is_unit (p.coeff k) :=
+begin
+  refine ⟨λ hp, ⟨hp.card_support_eq_three, λ k, hp.coeff_is_unit⟩, λ hp, _⟩,
+  obtain ⟨k, m, n, hkm, hmn, x, y, z, hx, hy, hz, rfl⟩ := card_support_eq_three.mp hp.1,
+  rw [support_trinomial hkm hmn hx hy hz] at hp,
+  replace hx := hp.2 k (mem_insert_self k {m, n}),
+  replace hy := hp.2 m (mem_insert_of_mem (mem_insert_self m {n})),
+  replace hz := hp.2 n (mem_insert_of_mem (mem_insert_of_mem (mem_singleton_self n))),
+  simp_rw [coeff_add, coeff_C_mul, coeff_X_pow_self, mul_one, coeff_X_pow] at hx hy hz,
+  rw [if_neg hkm.ne, if_neg (hkm.trans hmn).ne] at hx,
+  rw [if_neg hkm.ne', if_neg hmn.ne] at hy,
+  rw [if_neg (hkm.trans hmn).ne', if_neg hmn.ne'] at hz,
+  simp_rw [mul_zero, zero_add, add_zero] at hx hy hz,
+  exact ⟨k, m, n, hkm, hmn, x, y, z, hx, hy, hz, rfl⟩,
+end
+
+lemma is_unit_trinomial_iff' : p.is_unit_trinomial ↔ (p * p.mirror).coeff
+  (((p * p.mirror).nat_degree + (p * p.mirror).nat_trailing_degree) / 2) = 3 :=
+begin
+  rw [nat_degree_mul_mirror, nat_trailing_degree_mul_mirror, ←mul_add,
+      nat.mul_div_right _ zero_lt_two, coeff_mul_mirror],
+  refine ⟨_, λ hp, _⟩,
+  { rintros ⟨k, m, n, hkm, hmn, u, v, w, hu, hv, hw, rfl⟩,
+    rw [sum_def, support_trinomial hkm hmn hu.ne_zero hv.ne_zero hw.ne_zero,
+      sum_insert (mt mem_insert.mp (not_or hkm.ne (mt mem_singleton.mp (hkm.trans hmn).ne))),
+      sum_insert (mt mem_singleton.mp hmn.ne), sum_singleton, trinomial_leading_coeff' hkm hmn,
+      trinomial_middle_coeff hkm hmn, trinomial_trailing_coeff' hkm hmn],
+    simp only [hu, hv, hw, int.is_unit_sq, bit0, bit1, add_assoc] },
+  { have key : ∀ k ∈ p.support, (p.coeff k) ^ 2 = 1 :=
+    λ k hk, int.sq_eq_one_of_sq_le_three ((single_le_sum
+      (λ k hk, sq_nonneg (p.coeff k)) hk).trans hp.le) (mem_support_iff.mp hk),
+    refine is_unit_trinomial_iff.mpr ⟨_, λ k hk, is_unit_of_pow_eq_one _ 2 (key k hk) zero_lt_two⟩,
+    rw [sum_def, sum_congr rfl key, sum_const, nat.smul_one_eq_coe] at hp,
+    exact nat.cast_injective hp },
+end
+
+lemma is_unit_trinomial_iff'' (h : p * p.mirror = q * q.mirror) :
+  p.is_unit_trinomial ↔ q.is_unit_trinomial :=
+by rw [is_unit_trinomial_iff', is_unit_trinomial_iff', h]
+
+namespace is_unit_trinomial
+
+lemma irreducible_aux1 {k m n : ℕ} {u v w : ℤ}
+  (hkm : k < m) (hmn : m < n)
+  (hu : is_unit u) (hw : is_unit w)
+  (hp : p = C u * X ^ k + C v * X ^ m + C w * X ^ n) :
```

Comment:
Another way is to define ``local notation `p` := C u * X ^ k + C v * X ^ m + C w * X ^ n``, then you don't have to rewrite with `hp` so much.

Reply:
This gives some weird `«p»`.